File: app/src/main/java/org/breezyweather/main/widgets/LocationItemTouchCallback.java
package org.breezyweather.main.widgets;

import android.graphics.Canvas;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.Px;
import androidx.core.view.ViewCompat;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import org.breezyweather.common.basic.models.Location;
import org.breezyweather.common.ui.widgets.slidingItem.SlidingItemTouchCallback;
import org.breezyweather.main.MainActivityViewModel;
import org.breezyweather.R;
import org.breezyweather.common.basic.GeoActivity;
import org.breezyweather.common.utils.helpers.SnackbarHelper;

public class LocationItemTouchCallback extends SlidingItemTouchCallback {

    private final GeoActivity mActivity;
    private final MainActivityViewModel mViewModel;

    private final @Px int mElevation;

    private boolean mDragged;
    private int mDragFrom;
    private int mDragTo;
    private final @NonNull TouchReactor mReactor;

    public interface TouchReactor {
        void resetViewHolderAt(int position);
        void reorderByDrag(int from, int to);
        void startSelectProviderActivityBySwipe();
    }

    public LocationItemTouchCallback(GeoActivity activity, MainActivityViewModel viewModel,
                                     @NonNull TouchReactor callback) {
        super();

        mActivity = activity;
        mViewModel = viewModel;

        mElevation = activity.getResources().getDimensionPixelSize(R.dimen.touch_rise_z);

        mDragged = false;
        mReactor = callback;
    }

    @Override
    public void onSelectedChanged(@Nullable RecyclerView.ViewHolder viewHolder, int actionState) {
        super.onSelectedChanged(viewHolder, actionState);
        switch (actionState) {
            case ItemTouchHelper.ACTION_STATE_IDLE:
                if (mDragged) {
                    mDragged = false;
                    mViewModel.swapLocations(mDragFrom, mDragTo);
                }
                break;

            case ItemTouchHelper.ACTION_STATE_DRAG:
                if (!mDragged && viewHolder != null) {
                    mDragged = true;
                    mDragFrom = viewHolder.getBindingAdapterPosition();
                    mDragTo = mDragFrom;
                }
                break;
        }
    }

    @Override
    public boolean onMove(@NonNull RecyclerView recyclerView,
                          @NonNull RecyclerView.ViewHolder viewHolder,
                          @NonNull RecyclerView.ViewHolder target) {
        mDragTo = target.getBindingAdapterPosition();
        mReactor.reorderByDrag(viewHolder.getBindingAdapterPosition(), mDragTo);
        return true;
    }

    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
        int position = viewHolder.getBindingAdapterPosition();
        Location location = mViewModel.getTotalLocationList().getValue().getFirst().get(position);

        switch (direction) {
            case ItemTouchHelper.START: {
                if (location.isCurrentPosition()) {
                    mReactor.startSelectProviderActivityBySwipe();
                    mReactor.resetViewHolderAt(position);
                } else {
                    location = Location.copy(
                            location,
                            location.isCurrentPosition(),
                            !location.isResidentPosition()
                    );
                    mViewModel.updateLocation(location);

                    if (location.isResidentPosition()) {
                        SnackbarHelper.showSnackbar(
                                mActivity.getString(R.string.location_resident_message),
                                mActivity.getString(R.string.action_learn_more),
                                v -> new MaterialAlertDialogBuilder(mActivity)
                                        .setTitle(R.string.location_resident)
                                        .setMessage(R.string.location_resident_dialog)
                                        .show()
                        );
                    }
                }
                break;
            }
            case ItemTouchHelper.END:
                if (mViewModel.getTotalLocationList().getValue().getFirst().size() <= 1) {
                    // TODO: force update.
                    mViewModel.updateLocation(location);
                    SnackbarHelper.showSnackbar(
                            mActivity.getString(R.string.location_message_list_cannot_be_empty)
                    );
                } else {
                    location = mViewModel.deleteLocation(position);
                    SnackbarHelper.showSnackbar(
                            mActivity.getString(R.string.location_message_deleted),
                            mActivity.getString(R.string.action_cancel),
                            new CancelDeleteListener(location, position)
                    );
                }
                break;
        }
    }

    @Override
    public void onChildDraw(@NonNull Canvas c,
                            @NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder,
                            float dX, float dY, int actionState, boolean isCurrentlyActive) {
        super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        ViewCompat.setElevation(viewHolder.itemView,
                (dY != 0 || isCurrentlyActive) ? mElevation : 0);
    }

    // on click listener.

    private class CancelDeleteListener implements View.OnClickListener {

        private final Location mLocation;
        private final int mIndex;

        CancelDeleteListener(Location location, int index) {
            mLocation = location;
            mIndex = index;
        }

        @Override
        public void onClick(View view) {
            mViewModel.addLocation(mLocation, mIndex);
        }
    }
}

File: app/src/main/java/org/breezyweather/main/widgets/LocationItemTouchCallback.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.main.widgets

import android.graphics.Canvas
import android.view.View
import androidx.annotation.Px
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import breezyweather.domain.location.model.Location
import org.breezyweather.R
import org.breezyweather.common.basic.GeoActivity
import org.breezyweather.common.ui.widgets.slidingItem.SlidingItemTouchCallback
import org.breezyweather.common.utils.helpers.SnackbarHelper
import org.breezyweather.main.MainActivityViewModel

class LocationItemTouchCallback(
    private val mActivity: GeoActivity,
    private val mViewModel: MainActivityViewModel,
    private val mReactor: TouchReactor,
) : SlidingItemTouchCallback() {
    @Px
    private val mElevation: Int = mActivity.resources.getDimensionPixelSize(R.dimen.touch_rise_z)
    private var mDragged = false
    private var mDragFrom = 0
    private var mDragTo = 0

    interface TouchReactor {
        fun resetViewHolderAt(position: Int)
        fun reorderByDrag(from: Int, to: Int)
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        super.onSelectedChanged(viewHolder, actionState)
        when (actionState) {
            ItemTouchHelper.ACTION_STATE_IDLE -> if (mDragged) {
                mDragged = false
                mViewModel.swapLocations(mDragFrom, mDragTo)
            }

            ItemTouchHelper.ACTION_STATE_DRAG -> if (!mDragged && viewHolder != null) {
                mDragged = true
                mDragFrom = viewHolder.bindingAdapterPosition
                mDragTo = mDragFrom
            }
        }
    }

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder,
    ): Boolean {
        mDragTo = target.bindingAdapterPosition
        mReactor.reorderByDrag(viewHolder.bindingAdapterPosition, mDragTo)
        return true
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        val position = viewHolder.bindingAdapterPosition
        var location = mViewModel.validLocationList.value[position]
        when (direction) {
            ItemTouchHelper.START -> {
                viewHolder.bindingAdapter!!.notifyItemChanged(position)
                mViewModel.openChooseWeatherSourcesDialog(location)
            }

            ItemTouchHelper.END -> if (mViewModel.validLocationList.value.size <= 1) {
                viewHolder.bindingAdapter!!.notifyItemChanged(position)
                SnackbarHelper.showSnackbar(
                    mActivity.getString(R.string.location_message_list_cannot_be_empty)
                )
            } else {
                location = mViewModel.deleteLocation(position)
                SnackbarHelper.showSnackbar(
                    content = mActivity.getString(R.string.location_message_deleted),
                    action = mActivity.getString(R.string.action_undo),
                    listener = CancelDeleteListener(location, position)
                )
            }
        }
    }

    override fun onChildDraw(
        c: Canvas,
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        dX: Float,
        dY: Float,
        actionState: Int,
        isCurrentlyActive: Boolean,
    ) {
        super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
        ViewCompat.setElevation(
            viewHolder.itemView,
            (if (dY != 0f || isCurrentlyActive) mElevation else 0).toFloat()
        )
    }

    // on click listener.
    private inner class CancelDeleteListener(
        private val mLocation: Location,
        private val mIndex: Int,
    ) : View.OnClickListener {
        override fun onClick(view: View) {
            mViewModel.addLocation(mLocation, mIndex)
        }
    }
}
